fix: validate the presence of sortKeyFields when index is on the same field as the primary key

flozender opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#pull-requests
-->

#### Description of changes
Check if the LSI field is also a primary key and then make sure it has a `sortKeyFields` attribute. Appropriate unit test added.
<!--
Thank you for your Pull Request! Please provide a description above and review
the requirements below.
-->

#### Issue #, if available
Closes #4138

<!-- Also, please reference any associated PRs for documentation updates. -->

#### Description of how you validated changes
`yarn test`

#### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

flozender commented:
Looks like the test `owner-auth.test.ts` in `amplify-graphql-auth-transformer` has a field with an `index` on a `primaryKey` field with no `sortKeyFields`.

https://github.com/aws-amplify/amplify-cli/blob/fcf9100c778e98e16f0ed695112f2a7251aca06a/packages/amplify-graphql-auth-transformer/src/__tests__/owner-auth.test.ts#L241

What would you suggest I do here to fix this? Thanks!

cjihrig reviewed (changes requested):
@flozender you'll need to set the `secondarykeyasgsi` feature flag to `false` in your test, otherwise the Transformer will create a GSI by default. The validation logic will also need to check for that flag.

cjihrig commented on packages/amplify-graphql-index-transformer/src/graphql-index-transformer.ts:
```diff
@@ -88,6 +88,7 @@ export class IndexTransformer extends TransformerPluginBase {
 
 function validate(config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void {
   const { name, object, field, sortKeyFields } = config;
+  const defaultGSI = ctx.featureFlags.getBoolean('secondaryKeyAsGSI', false);
```
I think this should have a default value of `true`.

```suggestion
  const defaultGSI = ctx.featureFlags.getBoolean('secondaryKeyAsGSI', true);
```

cjihrig commented on packages/amplify-graphql-index-transformer/src/__tests__/amplify-graphql-index-transformer.test.ts:
```diff
@@ -52,6 +52,35 @@ test('throws if an invalid LSI is created', () => {
+        if (name === 'secondaryKeyAsGSI') {
+          return false;
+        }
+        return;
```
```suggestion
        return defaultValue;
```

cjihrig commented on packages/amplify-graphql-index-transformer/src/graphql-index-transformer.ts:
```diff
@@ -107,6 +108,19 @@ function validate(config: IndexDirectiveConfiguration, ctx: TransformerContextPr
 
     for (const peerDirective of objectField.directives!) {
+      if (
+        !defaultGSI &&
+        peerDirective.name.value === directiveName &&
+        objectField == config.primaryKeyField &&
+        !peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields')
```
This validation has some overlap with the validation on lines 131-134. Is it possible to do some consolidation.

flozender replied:
added `hasSortFields` to move the check into its own variable.

cjihrig reviewed (commented):
A few little comments, but this is looking good. Thanks.

flozender commented:
Thanks, I made the changes. Please let me know if something else seems off!

cjihrig commented on packages/amplify-graphql-index-transformer/src/graphql-index-transformer.ts:
```diff
@@ -107,17 +108,24 @@ function validate(config: IndexDirectiveConfiguration, ctx: TransformerContextPr
     for (const peerDirective of objectField.directives!) {
+      const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields');
+
+      if (!defaultGSI && !hasSortFields && peerDirective.name.value === directiveName && objectField == config.primaryKeyField) {
```
Does it work to move this validation down? The way this is currently written, it's possible that `config.primaryKeyField` won't be set yet. I'm also not sure that we need the `peerDirective.name.value === directiveName` part of this since we already know we are validating an `@index` directive.

flozender replied:
From what I understand, the `@index` shouldn't exist without a `sortKeyFields` when the `@index` is on the same field as a `@primaryKey`. So the only way `config.primaryKeyField` could be undefined would be if we had something like `foo: @index @primaryKey` would that be possible? Unfortunately, I cannot test it out myself since my AWS account is suspended right now.

cjihrig replied:
> `foo: @index @primaryKey` would that be possible?

Yes, I believe that would be possible.

> Unfortunately, I cannot test it out myself since my AWS account is suspended right now.

Since this only deals with input validation, you shouldn't need to deploy/test this in AWS.

flozender replied:
Thanks for the input, I made the changes and pushed it over again.

flozender replied:
Sorry, looks like moving it down fails quite a few tests. Do you think I should instead run a second `objectField.directives` loop for the validation after the current one so we know `config.primaryKeyField` will have to exist then?

flozender commented on packages/amplify-graphql-index-transformer/src/__tests__/amplify-graphql-index-transformer.test.ts:
```diff
@@ -52,6 +52,47 @@ test('throws if an invalid LSI is created', () => {
+      foo: ID!
+    }`;
+
+  const schemaInverted = `
```
added another test schema for the discussed case

sundersc commented on packages/amplify-graphql-index-transformer/src/graphql-index-transformer.ts:
```diff
@@ -112,12 +113,10 @@ function validate(config: IndexDirectiveConfiguration, ctx: TransformerContextPr
       }
 
       if (peerDirective.name.value === 'primaryKey') {
+        const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields');
```
Does it accommodate if sortKeyFields explicitly set as empty `sortKeyFields: []`?

flozender replied:
Yes, it does.

sundersc replied:
This schema failed to compile.
https://github.com/aws-amplify/amplify-cli/blob/abd8c67f7f4b76ece729f34447daf530179e1a1a/packages/amplify-util-mock/src/__e2e_v2__/model-relational-with-key-transformer.e2e.test.ts#L61-L66

flozender replied:
Thank you for your response, I was able to figure out what was going wrong. The tests should pass now.

sundersc commented:
Thanks for your contribution. The changes look good, except for the failing test.

sundersc reviewed (approved):
LGTM.

lazpavel reviewed (approved):
LGTM

phani-srikar reviewed (changes requested):
One unit test to cover a case. Otherwise LGTM

flozender commented:
Thanks for your comment @phani-srikar, I found that the code wasn't checking if `sortKeyFields` was empty if it was in the `index` directive. I've fixed that and added the test!

flozender commented on packages/amplify-graphql-index-transformer/src/graphql-index-transformer.ts:
```diff
@@ -24,6 +24,9 @@ const directiveDefinition = `
 
+/**
+ *
+ */
```
`eslint` force added this for `JSDocs`

flozender commented on packages/amplify-graphql-index-transformer/src/__tests__/amplify-graphql-index-transformer.test.ts:
```diff
@@ -43,14 +43,83 @@ test('throws if an invalid LSI is created', () => {
       id: ID! @primaryKey @index(name: "index1")
     }`;
 
+  const schemaEmptySortKeyFields = `
```
I found that this test case was missed for when using the `sortKeyFields` directive on `primaryKey`, so I added that test as well.

phani-srikar reviewed (approved):
Thanks for the changes. LGTM!
